homeassistant/components/mqtt/__init__.py

```diff
@@ -613,16 +629,20 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     mqtt_client.cleanup()
 
     # Trigger reload manual MQTT items at entry setup
-    # Reload the legacy yaml platform
-    await async_reload_integration_platforms(hass, DOMAIN, RELOADABLE_PLATFORMS)
     if (mqtt_entry_status := mqtt_config_entry_enabled(hass)) is False:
         # The entry is disabled reload legacy manual items when the entry is enabled again
         hass.data[DATA_MQTT_RELOAD_NEEDED] = True
     elif mqtt_entry_status is True:
         # The entry is reloaded:
         # Trigger re-fetching the yaml config at entry setup
         hass.data[DATA_MQTT_RELOAD_ENTRY] = True
-    # Stop the loop
+    # Reload the legacy yaml platform to make entities unavailable
+    await async_reload_integration_platforms(hass, DOMAIN, RELOADABLE_PLATFORMS)
```

Comment:
Why do we call this function here? Shouldn't it be called from `async_setup_entry`?

Reply:
We call it to reload all the reloadable platforms. The platforms will unsubscribe their MQTT topics and the entitities become unavailabe because they are not re-added until the entry is set up again.

Reply:
Entry unload is called at entry setup too. We remove the entities from the platform to be able to load a new config when the entry is started again. If the entry is disabled this will unload all entities from the platforms.